modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java

```diff
@@ -0,0 +1,272 @@
+
+    public final static Type TYPE = new Type("matrix_stats");
+    public final static AggregationStreams.Stream STREAM = new AggregationStreams.Stream() {
+        @Override
+        public InternalMatrixStats readResult(StreamInput in) throws IOException {
+            InternalMatrixStats result = new InternalMatrixStats();
+            result.readFrom(in);
+            return result;
+        }
+    };
+
+    public static void registerStreams() {
+        AggregationStreams.registerStream(STREAM, TYPE.stream());
+    }
+
+    /** per shard stats needed to compute stats */
+    protected RunningStats stats;
+    /** final result */
+    protected MatrixStatsResults results;
+
+    protected InternalMatrixStats() {
+    }
+
+    /** per shard ctor */
+    protected InternalMatrixStats(String name, long count, RunningStats multiFieldStatsResults, MatrixStatsResults results,
+                                  List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) {
+        super(name, pipelineAggregators, metaData);
+        assert count >= 0;
+        this.stats = multiFieldStatsResults;
+        this.results = results;
+    }
+
+    @Override
+    public Type type() {
+        return TYPE;
+    }
+
+    @Override
+    public long getDocCount() {
+        return stats.docCount;
+    }
+
+    @Override
+    public long getFieldCount(String field) {
+        if (results == null) {
+            return 0;
+        }
+        return results.getFieldCount(field);
+    }
+
+    @Override
+    public Double getMean(String field) {
+        if (results == null) {
+            return null;
+        }
+        return results.getMean(field);
+    }
+
+    @Override
+    public Double getVariance(String field) {
+        if (results == null) {
+            return null;
+        }
+        return results.getVariance(field);
+    }
+
+    @Override
+    public Double getSkewness(String field) {
+        if (results == null) {
+            return null;
+        }
+        return results.getSkewness(field);
+    }
+
+    @Override
+    public Double getKurtosis(String field) {
+        if (results == null) {
+            return null;
+        }
+        return results.getKurtosis(field);
+    }
+
+    @Override
+    public Double getCovariance(String fieldX, String fieldY) {
+        if (results == null) {
+            return null;
+        }
+        return results.getCovariance(fieldX, fieldY);
+    }
+
+    @Override
+    public Map<String, HashMap<String, Double>> getCovariance() {
+        return results.getCovariances();
+    }
+
+    @Override
+    public Double getCorrelation(String fieldX, String fieldY) {
+        if (results == null) {
+            return null;
+        }
+        return results.getCorrelation(fieldX, fieldY);
+    }
+
+    @Override
+    public Map<String, HashMap<String, Double>> getCorrelation() {
+        return results.getCorrelations();
+    }
```

Comment:
in terms of API I think we should choose which one of the two above methods we want to expose rather than exposing both